// snack-sdk/build/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.parseRuntimeUrl = exports.createRuntimeUrl = exports.Snack = exports.defaultConfig = exports.getDeprecatedModule = exports.standardizeDependencies = exports.isFeatureSupported = exports.getSupportedSDKVersions = exports.isValidSemver = exports.getPreloadedModules = exports.isModulePreloaded = void 0;
var snack_content_1 = require("snack-content");
Object.defineProperty(exports, "isModulePreloaded", { enumerable: true, get: function () { return snack_content_1.isModulePreloaded; } });
Object.defineProperty(exports, "getPreloadedModules", { enumerable: true, get: function () { return snack_content_1.getPreloadedModules; } });
Object.defineProperty(exports, "isValidSemver", { enumerable: true, get: function () { return snack_content_1.isValidSemver; } });
Object.defineProperty(exports, "getSupportedSDKVersions", { enumerable: true, get: function () { return snack_content_1.getSupportedSDKVersions; } });
Object.defineProperty(exports, "isFeatureSupported", { enumerable: true, get: function () { return snack_content_1.isFeatureSupported; } });
Object.defineProperty(exports, "standardizeDependencies", { enumerable: true, get: function () { return snack_content_1.standardizeDependencies; } });
Object.defineProperty(exports, "getDeprecatedModule", { enumerable: true, get: function () { return snack_content_1.getDeprecatedModule; } });
Object.defineProperty(exports, "createRuntimeUrl", { enumerable: true, get: function () { return snack_content_1.createRuntimeUrl; } });
Object.defineProperty(exports, "parseRuntimeUrl", { enumerable: true, get: function () { return snack_content_1.parseRuntimeUrl; } });
var Session_1 = __importDefault(require("./Session"));
exports.Snack = Session_1.default;
var defaultConfig_1 = __importDefault(require("./defaultConfig"));
exports.defaultConfig = defaultConfig_1.default;
__exportStar(require("./transports"), exports);
__exportStar(require("./types"), exports);
